[PM-18800] vault onboarding nudges and badge

Jingo88 opened the pull request:
## 🎟️ Tracking

[PM-18800](https://bitwarden.atlassian.net/browse/PM-18800)

## 📔 Objective

* Add Import Nudge to Empty Browser Vault
* Add Welcome Nudge to New Account (within 30 days) populated Vault
* Add a badge to `Vault options` inside the settings tab
* Update copy for empty vault
* Update copy for vault inside settings tab

NOTES
* Moved the config service flag calls from the components to the services to reduce `code churn`. This can be seen in the IntroCarousel as well as the Vault Nudges. 
* Updated the Vault Nudges Service to use a new type `NudgeStatus` that will hold the dismiss value of the `spotlight` and the `badge` separately. This will allow components to have better control over dismissing a `badge` without impacting the `spotlight`

## 📸 Screen Recordings

* Empty Vault Nudge Dismissal

https://github.com/user-attachments/assets/0af86b55-8801-42a9-a597-3a5737672382


* Empty Vault Add Item

https://github.com/user-attachments/assets/81d92770-9503-438e-810e-17a78569accf

* Settings Vault Badge Dismissal


https://github.com/user-attachments/assets/7146b483-c89b-416b-9462-b26549734546


* Settings Vault Badge Dismissed By Nudge


https://github.com/user-attachments/assets/13052044-4b3c-4a82-9289-51fe7f2dec5d


* Has Items Vault Welcome Nudge Dismiss

https://github.com/user-attachments/assets/8780a89c-e841-464e-a449-c3a566aaf55c







## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

<!-- Suggested interactions but feel free to use (or not) as you desire! -->

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes


[PM-18800]: https://bitwarden.atlassian.net/browse/PM-18800?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

audreyality commented on apps/browser/src/tools/popup/settings/settings-v2.component.ts:
```diff
@@ -22,6 +29,38 @@ import { PopupPageComponent } from "../../../platform/popup/layout/popup-page.co
+      FeatureFlag.PM8851_BrowserOnboardingNudge,
+    );
+    if (this.hasVaultNudgeFlag) {
+      this.showVaultBadge$ = combineLatest([
+        this.vaultNudgesService.showNudge$(VaultNudgeType.EmptyVaultNudge, this.activeUserId),
+        this.vaultNudgesService.showNudge$(VaultNudgeType.EmptyVaultBadge, this.activeUserId),
+      ]).pipe(
+        switchMap(([showEmptyVaultNudge, showEmptyVaultBadge]) => {
+          return of(showEmptyVaultNudge && showEmptyVaultBadge);
+        }),
+      );
+    }
+  }
+
+  async dismissBadge(type: VaultNudgeType) {
+    await this.vaultNudgesService.dismissNudge(type, this.activeUserId as UserId);
```
🤔 It seems odd to have a badge and a nudge when they're so closely coupled together. Is there a reason why the distinction can't be encapsulated by the service?

audreyality commented on apps/browser/src/tools/popup/settings/settings-v2.component.ts:
```diff
@@ -22,6 +29,38 @@ import { PopupPageComponent } from "../../../platform/popup/layout/popup-page.co
+  showVaultBadge$: Observable<boolean> = new Observable();
+  activeUserId: UserId | null = null;
+
+  constructor(
+    private readonly vaultNudgesService: VaultNudgesService,
+    private readonly accountService: AccountService,
+    private readonly configService: ConfigService,
+  ) {}
+  async ngOnInit() {
+    this.activeUserId = await firstValueFrom(this.accountService.activeAccount$.pipe(getUserId));
+    this.hasVaultNudgeFlag = await this.configService.getFeatureFlag(
+      FeatureFlag.PM8851_BrowserOnboardingNudge,
+    );
```
💭 I've been thinking about the way that we use the `configService` and incremental enhancements on the tools team, where we've traditionally injected the `configService` in order to control interface logic, as you're doing here. The key insight for me is that wherever `ConfigService` is, it'll end up being removed once the flag is no longer useful... it introduces **code churn**.

🎨 From there I reason about whether the system can be designed to manage this churn throug dependency injection. It seems like this may be such a case. Your `VaultNudgeService` is brand new. You could create two versions: A stub used when the flag is disabled and the other when the flag is enabled. That, then, allows you to hoist feature flag decisions into a factory function.

audreyality reviewed (dismissed):
👍🏻 Everything on the tools side looks good. I have some comments on your use of feature flags, but whether or not you attempt to change your design is totally up to you.

❓ Is this only going to be a browser feature? It seems that we'd want it on the other clients, which would mean constructing these components in a library and only adding host components to the browser.

Jingo88 commented:
Thanks for the comments @audreyality. Gonna convert to draft for now and noodle on this a bit.

Jingo88 commented:
I refactored the service a bit. 

1. Created a `NudgeStatus` type to address the difference between a `spotlight` call to action, vs the `badge` value (going to use `nudge` term as the overall status). This logic is now kept in the service and the component can better differentiate when one or the other is needed.
2. I moved the ConfigService flag call into the service themselves. (VaultNudgeService && IntroCarouselService). This should keep the feature flag calls consolidated, future components will not need to worry about importing and calling configService.

Note: I attempted the stub and factory idea but I'm not sure how to provide the factory in a synchronous way to wait for the flag call. I believe consolidating the flag call into the service methods themselves should address the code churn. 

Thanks @audreyality for the input, and thanks to @shane-melton for the help.

audreyality reviewed (approved):
👍🏻 Much improved! It's easier for me to follow the code!

🌱 I'll have to take a closer look at angular DI and see if I can solve your async issue... but not for this PR. 😄

shane-melton commented on apps/browser/src/vault/popup/components/vault-v2/vault-v2.component.html:
```diff
@@ -8,20 +8,50 @@
 
+  <ng-container
+    slot="above-scroll-area"
+    *ngIf="vaultState !== VaultStateEnum.Empty && (showHasItemsVaultNudge$ | async)?.showSpotlight"
```
🎨 This is going to conflict with the other `above-scroll-area` slots in the template. I think we need to wrap all of these in a singular `<ng-container slot="above-scroll-area" />` and have the conditionals within for each spotlight, header, etc.

shane-melton commented on libs/vault/src/services/custom-nudges-services/empty-vault-nudge.service.ts:
```diff
@@ -0,0 +1,68 @@
+        });
+        const managedCollections = collections.filter((collection) => {
+          if (orgIds.has(collection.organizationId) && collection.manage) {
+            return of(true);
```
⛏️  Returning `of(true)` inside `Array.filter` isn't the right type. I think you can just return the condition in the `if`.

shane-melton commented on libs/vault/src/services/custom-nudges-services/empty-vault-nudge.service.ts:
```diff
@@ -0,0 +1,68 @@
+        const canCreateCollections = orgs.filter((org) => {
+          return org.canCreateNewCollections;
+        });
+        const managedCollections = collections.filter((collection) => {
```
🎨 I think this is a little easier to read and we don't have to iterate through all the collections after finding one with `manage==true`.
```ts
const hasManageCollections = collections.some((c) => c.manage && orgIds.has(c.organizationId));
```

shane-melton commented on libs/vault/src/services/custom-nudges-services/has-items-nudge.service.ts:
```diff
@@ -1,30 +1,46 @@
-      ),
+  shouldShowNudge$(nudgeType: VaultNudgeType, userId: UserId): Observable<NudgeStatus> {
+    return combineLatest([
+      this.apiService.getProfile(),
```
💭 We should be mindful of how this behaves when the application is offline since we're making an API request with `getProfile()` that will fail.

shane-melton replied:
@Jingo88 The vault profile service can still potentially throw an exception if the client is offline, we should probably still wrap it in a try/catch and handle that scenario gracefully.

shane-melton commented on libs/vault/src/services/vault-nudges.service.ts:
```diff
@@ -1,30 +1,37 @@
 import { DefaultSingleNudgeService, SingleNudgeService } from "./default-single-nudge.service";
 
+export type NudgeStatus = {
+  showBadge: boolean;
```
❓ 💭 Are we sure we want these to be `show` flags? I know before this change, the value stored in state represented a "dismissed" state. Meaning if it was missing from state we would default to show the nudge. Now we need to store a default value of `true` in order to get the nudge to show.

I don't feel too strongly either way, only wanted to make sure we considered both.

shane-melton commented on libs/vault/src/services/custom-nudges-services/has-nudge.service.ts:
♻️ (non-blocking): I missed when this service was added, but it feels like it should've been part of the `VaultNudgeService` directly as an new method. Something like `VaultNudgeService.hasActiveNudges$(userId)` instead of an entirely separate service.

It's also misusing the `DefaultSingleNudgeService` abstract class, because this is an entirely separate concept (does the user have any nudge active vs. does the user have a single, particular nudge).

I don't think we necessarily have to change this here, and maybe you/others feel differently, but I wanted to bring it up.

Jingo88 replied:
I see what you mean. I'm going to leave this as is for now. But I'll add a refactor ticket once this is merged.

shane-melton replied:
Sounds good, thanks!

shane-melton commented on apps/browser/src/vault/popup/components/vault-v2/vault-v2.component.ts:
```diff
@@ -74,14 +84,25 @@ enum VaultState {
+  VaultNudgeType = VaultNudgeType;
   cipherType = CipherType;
-
+  showEmptyVaultNudge$: Observable<NudgeStatus> = of({
```
🎨 (non-blocking): It seems like a code smell that we're creating these placeholder observables that are immediately replaced in `ngOnInit`.

I think it would be cleaner to just define them outright at the top level instead of overwriting them in `ngOnInit`. Something like:

```ts
private activeUserId$ = this.accountService.activeAccount$.pipe(getUserId);
showEmptyVaultNudge$: Observable<boolean> = this.activeUserId$.pipe(
  switchMap((userId) => this.vaultNudgesService.showNudge$(VaultNudgeType.EmptyVaultNudge, userId)),
  map((nudgeStatus) => !nudgeStatus.hasSpotlightDismissed), // Also can map to boolean to avoid the `?.` in the template
);
```

shane-melton commented on apps/browser/src/vault/popup/services/intro-carousel.service.ts:
```diff
@@ -26,9 +28,17 @@ export class IntroCarouselService {
 
   async setIntroCarouselDismissed(): Promise<void> {
-    await this.introCarouselState.update(() => true);
+    const hasVaultNudgeFlag = await firstValueFrom(
```
❓ Why do we need to check the flag here again? The guard should prevent the user from getting to the carousel in the first place if the flag is disabled, correct?

Jingo88 replied:
The guard is only on the login route. This flag call was previously in the vault-v2.component to cover users who were already logged in. I just moved it to the service from our discussion about consolidating flag calls to services.

shane-melton reviewed (commented):
Great work on this and the refactor! Mostly had some non-blocking questions/suggestions.

Jingo88 commented on libs/vault/src/services/vault-nudges.service.ts:
```diff
@@ -56,15 +65,30 @@ export class VaultNudgesService {
-  dismissNudge(nudge: VaultNudgeType, userId: UserId) {
-    return this.getNudgeService(nudge).setNudgeStatus(nudge, true, userId);
+  async dismissNudge(nudge: VaultNudgeType, userId: UserId, onlyBadge: boolean = false) {
+    const hasVaultNudgeFlag = await this.configService.getFeatureFlag(
```
Ah yeah, that's true, it shouldn't be available unless the showNudge$ returned true. Removed!

shane-melton reviewed (dismissed):
🚀 Nice work!

audreyality commented on apps/browser/src/popup/tabs-v2.component.ts:
```diff
@@ -18,7 +18,7 @@ export class TabsV2Component {
   protected navButtons$ = combineLatest([
     this.configService.getFeatureFlag$(FeatureFlag.PM8851_BrowserOnboardingNudge),
-    this.hasNudgeService.shouldShowNudge$(),
+    this.hasNudgeService.nudgeStatus$(),
```
⛏️ You should also update the `showNudge` binding on line 23.

audreyality reviewed (changes requested):
Just a nit.
